chihuahua/src/hcall/common.rs

```diff
@@ -441,6 +455,25 @@ impl<Module, Memory> HostProvisioningState<Module, Memory> {
         Some(&self.data_sources[index])
     }
 
+    /// Returns the data source frame (containing raw data, source and package
+    /// IDs) associated with the Nth registered data source as stored in the
+    /// host provisioning state, expressed by `index`.  Returns `None` iff
+    /// `index` is not the index of any registered data source.
+    pub(crate) fn get_current_stream_source(&self, index: usize) -> Option<&DataSourceMetadata> {
+        if index > self.get_current_stream_source_count() {
+            return None;
+        }
+        Some(&self.stream_sources[index])
+    }
+
+    /// Returns the data source frame (containing raw data, source and package
+    /// IDs) associated with the Nth registered data source as stored in the
+    /// host provisioning state, expressed by `index`.  Returns `None` iff
+    /// `index` is not the index of any registered data source.
+    pub(crate) fn get_previous_result(&self) -> Option<Vec<u8>> {
+        self.previous_result.clone()
```

Comment:
I don't think this should `clone`.  Rather it should return `Option<&Vec<u8>>` and if the caller needs to clone the result then they should do it.